src/fparser/statements.py

```diff
@@ -1748,47 +1750,169 @@ class Case(Statement):
+        if self.name and self.name != parent_name:
+            self.warning(
+                'expected type-is-construct-name %r but got %r, skipping.'
+                % (parent_name, self.name))
+            self.isvalid = False
+        return
+
+    def tofortran(self, isfix=None):
+        ''' Create the Fortran representation of this object and return
+        it as a string '''
+        tab = self.get_indent_tab(isfix=isfix)
+        text = tab + 'TYPE IS'
+        if self.items:
+            lst = []
+            for item in self.items:
+                lst.append((' : '.join(item)).strip())
+            text += ' ( %s )' % (', '.join(lst))
+        else:
+            raise ParseError("TYPE IS construct must have arguments")
+        if self.name:
+            text += ' ' + self.name
+        return text
+
+    def analyze(self):
+        return
+
+
+class ClassIs(Statement):
+    """
+    CLASS <class-selector>
+    <class-selector> = ( IS <type-value-range-list> | DEFAULT )
+    <type-value-range> = <case-value>
+    <case-value> = <char>
+    """
+    match = re.compile(r'class\b\s*(is\b\s*\(.*\)|default)\s*\w*\Z',
+                       re.I).match
+
+    def process_item(self):
+        ''' Populate the state of this object by parsing the string '''
+        line = self.item.get_line()[5:].lstrip()
+        if line.lower().startswith('is'):
+            # We have a 'class is ...' statement
+            idx = line.find('(')
+            line = line[idx:]
+            idx = line.find(')')
+            items = split_comma(line[1:idx].strip(), self.item)
+            line = line[idx+1:].lstrip()
+        else:
+            # We have a 'class default' statement
+            assert line.lower().startswith('default'), repr(line)
+            items = []
+            line = line[7:].lstrip()
+        for idx in range(len(items)):
+            item = self.item.copy(items[idx])
+            rlst = []
+            for rchar in item.get_line().split(':'):
+                rlst.append(item.apply_map(rchar.strip()))
+            items[idx] = rlst
+        self.items = items
+        self.name = line
+        parent_name = getattr(self.parent, 'name', '')
+        if self.name and self.name != parent_name:
+            self.warning(
+                'expected class-construct-name %r but got %r, skipping.'
+                % (parent_name, self.name))
+            self.isvalid = False
+        return
+
+    def tofortran(self, isfix=None):
+        ''' Returns the Fortran for this object as a string '''
+        tab = self.get_indent_tab(isfix=isfix)
+        text = tab + 'CLASS'
+        if self.items:
+            text += ' IS'
+            lchar = []
+            for item in self.items:
+                lchar.append((' : '.join(item)).strip())
+            text += ' ( %s )' % (', '.join(lchar))
+        else:
+            text += ' DEFAULT'
+        if self.name:
+            text += ' ' + self.name
```

Comment:
The SELECT block can (optionally) have a name. This name must then be appended to each clause of the parent block. However, as noted earlier, the parser actually barfs if you attempt to name a SELECT (#21).